filters/tee/tee_test.go

```diff
@@ -0,0 +1,129 @@
+)
+
+func TestTeeHostHeaderChanges(t *testing.T) {
+	f, _ := testTeeSpec.CreateFilter(teeArgsAsBackend)
+	fc := buildfilterContext()
+
+	rep, _ := f.(*tee)
+	modifiedRequest := cloneRequest(rep, fc.Request())
+	if modifiedRequest.Host != "api.example.com" {
+		t.Error("Tee Request Host not modified")
+	}
+
+	originalRequest := fc.Request()
+	if originalRequest.Host == "api.example.com" {
+		t.Error("Incoming Request Host was modified")
+	}
+}
+
+func TestTeeSchemeChanges(t *testing.T) {
+	f, _ := testTeeSpec.CreateFilter(teeArgsAsBackend)
+	fc := buildfilterContext()
+
+	rep, _ := f.(*tee)
+	modifiedRequest := cloneRequest(rep, fc.Request())
+	if modifiedRequest.URL.Scheme != "https" {
+		t.Error("Tee Request Scheme not modified")
+	}
+
+	originalRequest := fc.Request()
+	if originalRequest.URL.Scheme == "https" {
+		t.Error("Incoming Request Scheme was modified")
+	}
+}
+
+func TestTeeUrlHostChanges(t *testing.T) {
+	f, _ := testTeeSpec.CreateFilter(teeArgsAsBackend)
+	fc := buildfilterContext()
+
+	rep, _ := f.(*tee)
+	modifiedRequest := cloneRequest(rep, fc.Request())
+	if modifiedRequest.URL.Host != "api.example.com" {
+		t.Error("Tee Request Url Host not modified")
+	}
+
+	originalRequest := fc.Request()
+	if originalRequest.URL.Host == "api.example.com" {
+		t.Error("Incoming Request Url Host was modified")
+	}
+}
+
+func TestTeeWithPathChanges(t *testing.T) {
+	f, _ := testTeeSpec.CreateFilter(teeArgsWithModPath)
+	fc := buildfilterContext()
+
+	rep, _ := f.(*tee)
+	modifiedRequest := cloneRequest(rep, fc.Request())
+	if modifiedRequest.URL.Path != "/v1/" {
+		t.Errorf("Tee Request Path not modified, %v", modifiedRequest.URL.Path)
+	}
+
+	originalRequest := fc.Request()
+	if originalRequest.URL.Path != "/api/v3" {
+		t.Errorf("Incoming Request Scheme modified, %v", originalRequest.URL.Path)
+	}
+}
+
+type MyHandler struct {
+	name string
+	body string
+}
+
+func (h *MyHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	b, _ := ioutil.ReadAll(r.Body)
+	str := string(b)
+	h.body = str
+}
+
+func TestTeeEndToEndBody(t *testing.T) {
+	shadowHandler := &MyHandler{name: "shadow"}
+	shadowServer := httptest.NewServer(shadowHandler)
+	shadowUrl := shadowServer.URL
+	defer shadowServer.Close()
+
+	originalHandler := &MyHandler{name: "original"}
+	originalServer := httptest.NewServer(originalHandler)
+	originalUrl := originalServer.URL
+	defer originalServer.Close()
+
+	routeStr := "route1: * -> Tee(\"" + shadowUrl + "\")" + " -> " + "\"" + originalUrl + "\";"
+
+	route, _ := eskip.Parse(routeStr)
+	registery := make(filters.Registry)
+	registery.Register(NewTee())
+	p := proxytest.New(registery, route...)
+	defer p.Close()
+
+	req, _ := http.NewRequest("GET", p.URL, strings.NewReader("TESTEST"))
+	req.Host = "www.example.org"
+	req.Header.Set("X-Test", "true")
+	req.Close = true
+	rsp, _ := (&http.Client{}).Do(req)
+	rsp.Body.Close()
+	if shadowHandler.body != originalHandler.body {
```

Comment:
the problem with this test is that, in case of some lower level failure, both the shadowHandler and the originalHandler can have empty body, and the test may give false positive. Either make the test fail in case of error at line 119, or check if the two bodies are actually equal to "TESTTEST", or both (best ;) )